dev/io.openliberty.microprofile.faulttolerance.telemetry/src/io/openliberty/microprofile/faulttolerance/telemetry/metrics/integration/AbstractMetricRecorderImpl.java

```diff
@@ -0,0 +1,474 @@
+            ftCircuitbreakerStateTotal = meter.counterBuilder(metricPrefix + ".circuitbreaker.state.total").setUnit("nanoseconds").setDescription(Tr.formatMessage(tc, "ft.circuitbreaker.state.total.description")).buildWithCallback(this::getCircuitBreakerAccumulatedTimes);
+            ftCircuitbreakerOpenedTotal = meter.counterBuilder(metricPrefix + ".circuitbreaker.opened.total").setDescription(Tr.formatMessage(tc, "ft.circuitbreaker.opened.total.description")).build();
+        } else {
+            ftCircuitbreakerCallsTotal = null;
+            ftCircuitbreakerStateTotal = null;
+            ftCircuitbreakerOpenedTotal = null;
+        }
+
+        if (bulkheadPolicy != null) {
+            ftBulkheadCallsTotal = meter.counterBuilder(metricPrefix + ".bulkhead.calls.total").setDescription(Tr.formatMessage(tc, "ft.bulkhead.calls.total.description")).build();
+            ftBulkheadExecutionsRunning = meter.gaugeBuilder(metricPrefix + ".bulkhead.executionsRunning").ofLongs().setDescription(Tr.formatMessage(tc, "ft.bulkhead.executionsRunning.description")).buildWithCallback(this::getConcurrentExecutions);
+            ftBulkheadRunningDuration = meter.histogramBuilder(metricPrefix + ".bulkhead.runningDuration").setDescription(Tr.formatMessage(tc, "ft.bulkhead.runningDuration.description")).ofLongs().setUnit("nanoseconds").build();
+
+            if (isAsync == AsyncType.ASYNC) {
+                ftBulkheadExecutionsWaiting = meter.gaugeBuilder(metricPrefix + ".bulkhead.executionsWaiting").ofLongs().setDescription(Tr.formatMessage(tc, "ft.bulkhead.executionsWaiting.description")).buildWithCallback(this::getQueuePopulation);
+                ftBulkheadWaitingDuration = meter.histogramBuilder(metricPrefix + ".bulkhead.waitingDuration").setDescription(Tr.formatMessage(tc, "ft.bulkhead.waitingDuration.description")).ofLongs().setUnit("nanoseconds").build();
+            } else {
+                ftBulkheadExecutionsWaiting = null;
+                ftBulkheadWaitingDuration = null;
+            }
+        } else {
+            ftBulkheadCallsTotal = null;
+            ftBulkheadExecutionsRunning = null;
+            ftBulkheadExecutionsWaiting = null;
+            ftBulkheadRunningDuration = null;
+            ftBulkheadWaitingDuration = null;
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Trivial
+    @Override
+    public void incrementInvocationSuccessCount(FallbackOccurred fallbackOccurred) {
+        incrementFtTotalCount(fallbackOccurred, true);
+    }
+
+    /** {@inheritDoc} */
+    @Trivial
+    @Override
+    public void incrementInvocationFailedCount(FallbackOccurred fallbackOccurred) {
+        incrementFtTotalCount(fallbackOccurred, false);
+    }
+
+    @Trivial
+    private void incrementFtTotalCount(FallbackOccurred fallbackOccurred, boolean invocationSucceeded) {
+        AttributesBuilder ab = Attributes.builder().putAll(methodAttribute);
+
+        if (!fallbackDefined) {
+            ab.put("fallback", "notDefined");
```

Comment:
Where we're building an `Attributes`, we should be using a pre-declared `AttributeKey<T>` as the key, based on the note on [put(String, String)](https://www.javadoc.io/static/io.opentelemetry/opentelemetry-api/1.38.0/io/opentelemetry/api/common/AttributesBuilder.html#put(java.lang.String,java.lang.String))

That would mean declaring an `AttributeKey<String>` constant for each key.